_posts/2023-01-06-Algorithms-Graphs-BFS-DFS.md

```diff
@@ -0,0 +1,329 @@
+
+| 0    | 1    | 1    | 0    | 0    |
+| ---- | ---- | ---- | ---- | ---- |
+| 1    | 0    | 0    | 0    | 1    |
+| 1    | 0    | 0    | 1    | 0    |
+| 0    | 1    | 1    | 1    | 0    |
+| 0    | 1    | 0    | 0    | 0    |
+
+This matrix shows that element node 1 (columns 1) is connected with node 2 and 3 (so that's why row 2 and row 4 has ones in column 1).
+
+Then node 2 (column 2) is connected with node 1, 4, 5 (that's why in row 1, 4 and 5 there is "1" in column 2), etc.
+
+To be honest I have never used matrix representation in any graph related problem, but for sure it is good to know that this kind of representation exists.
+
+### Very popular graph terms
+
+- **Cycle**: if within a graph we travel and meet the visited node once again: it means that we have a cycle in the graph. Finding a path of a cycle can be complicated and we can use an algorithm to do that (BFS or DFS - more about them in further parts of the article).
+
+<p align="center">
+<img src="https://raw.githubusercontent.com/swiftingio/blog/%2357-Algorithms-DFS-BFS/Cycle.svg" width="643">
+</p>
+
+- **Connected component (cc)**: we can imagine a connected component as a lonely island which is not connected to other lands. There exist algorithms for finding connected components in the graph. I hope in the future I will write more about it in a separate post.
+
+<p align="center">
+<img src="https://raw.githubusercontent.com/swiftingio/blog/%2357-Algorithms-DFS-BFS/ConnectedComponents.svg" width="643">
+</p>
+
+### The most popular graph traversal algorithm
+
+Let's focus on the main subject of this post: Graph traversal algorithms. We can distinguish two:
+
+- DFS (Depth First Search)
+- BFS (Breadth First Search)
+
+It is really important to understand how they work and when they are most useful (in which use case). Below I tried to highlight the main features of both. I presented it in the table for a better understanding of the difference between them.
+
+|                                                      | DFS                                                          | BFS                                                          |
+| :--------------------------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| **Main usability**                                   | Is commonly used for visiting (traversing) all nodes in the graph (or to be more concrete: in connected component ) | Is used when we want to find the shortest path in the unweighted graph |
+| **Implementation**                                   | Using recursion (iterative as well but definitely less popular and used) | Iterative                                                    |
+| **Data structure used for iterative implementation** | Stack                                                        | Queue                                                        |
```

Comment:
Tutaj piszesz że dla DFS jest zastosowany stos, a w implementacji jest Set. Może warto zmienić implementację na Array (nowe elementy są dodawane do końca tablicy, tak jakby był to stos). Dzięki tablicy można też odtworzyć kolejność, w jakiej algorytm odwiedzał wszystkie węzły (nodes).

Reply:
Jest zastosowany stos rekurencyjny. Set jest wykorzystywany do sprawdzenia czy widzieliśmy już element, żeby nie powtarzać tych samych nodów